Blog for local storage capacity isolation GA

jingxu97 opened the pull request:
blog for GA feature local storage capacity isolation

<!-- ℹ️

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

jingxu97 commented:
/assign @xing-yang

AnaMMedina21 commented:
Hi from Comms team  👋🏽
Thanks for getting the blog PR in! 
We are planning to publish your blog post on September 19, 2022. Let us know if you have any questions

divya-mohan0209 commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+**Author:** Jing Xu (Google)
+
+Local ephemeral storage capacity isolation was introduced as a alpha feature in Kubernetes 1.7 and it went beta in 1.9 and with Kubernetes 1.25 we are excited to announce general availability(GA) of this feature.
```
```suggestion
Local ephemeral storage capacity isolation was introduced as an alpha feature in Kubernetes 1.7 and it went beta in 1.9. With Kubernetes 1.25, we are excited to announce the general availability (GA) of this feature.
```

jingxu97 replied:
done

divya-mohan0209 commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+Local ephemeral storage capacity isolation was introduced as a alpha feature in Kubernetes 1.7 and it went beta in 1.9 and with Kubernetes 1.25 we are excited to announce general availability(GA) of this feature.
+
+Pods use ephemeral local storage for scratch space, caching, and for logs. The lifetime of local ephemeral storage does not extend beyond the life of the individual pod. It is exposed to pods using the container’s writable layer, logs directory, and EmptyDir volumes. Before this feature available, there have been issues related to the lack of local storage accounting and isolation, such as Pods d ...
```
I'm not sure if EmptyDir needs to be encapsulated in a codeblock. I did it. 
```suggestion
Pods use ephemeral local storage for scratch space, caching, and logs. The lifetime of local ephemeral storage does not extend beyond the life of the individual pod. It is exposed to pods using the container’s writable layer, logs directory, and `EmptyDir` volumes. Before this feature was introduced, there were issues related to the lack of local storage accounting and isolation, such as Pods not knowing how much local storage is available and being unable to request guaranteed local storage. Local storage is a best-effort resource and pods can be evicted due to other pods filling the local storage.
```

jingxu97 replied:
updated

divya-mohan0209 commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+Pods use ephemeral local storage for scratch space, caching, and for logs. The lifetime of local ephemeral storage does not extend beyond the life of the individual pod. It is exposed to pods using the container’s writable layer, logs directory, and EmptyDir volumes. Before this feature available, there have been issues related to the lack of local storage accounting and isolation, such as Pods d ...
+
+The local storage capacity isolation feature allows users to manage local ephemeral storage in the same way as managing CPU and memory. It provides support for capacity isolation of shared storage between pods, such that a pod can be hard limited in its consumption of shared resources by evicting Pods if its consumption of shared storage exceeds that limit. It also allows setting ephemeral storag ...
```
We should link the feature to the concepts docs, if possible or if we have a feature doc, even better.

```suggestion
The [local storage capacity isolation feature](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage) allows users to manage local ephemeral storage in the same way as managing CPU and memory. It provides support for capacity isolation of shared storage between pods, such that a pod can be hard limited in its consumption of shared resources by evicting Pods if its consumption of shared storage exceeds that limit. It also allows setting ephemeral storage requests for resource reservation. The limits and requests for shared `ephemeral-storage` are similar to those for memory and CPU consumption.
```

divya-mohan0209 commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+### How to use local storage capacity isolation
+
+A typical configuration for local ephemeral storage is to place all different kinds of ephemeral local data (emptyDir volumes, writeable layers, container images, logs) into one filesystem. Typically, both /var/lib/kubelet and /var/log are on the system root filesystem. If users configure the local storage in different ways, kubelet might not be able to correctly measure disk usage and use this f ...
```
```suggestion
A typical configuration for local ephemeral storage is to place all different kinds of ephemeral local data (emptyDir volumes, writeable layers, container images, logs) into one filesystem. Typically, both /var/lib/kubelet and /var/log are on the system's root filesystem. If users configure the local storage in different ways, kubelet might not be able to correctly measure disk usage and use this feature.
```

divya-mohan0209 commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+* `spec.containers[].resources.limits.ephemeral-storage`
+* `spec.containers[].resources.requests.ephemeral-storage`
+
+In the following example, the Pod has two containers. The first container has a request of 2GiB of local ephemeral storage and a limit of 5GiB. The second container requests 6GiB of local storage. Therefore, the Pod has a request of total 10GiB (8GiB+2GiB) of local ephemeral storage, and a limit of 6GiB of local ephemeral storage. It also sets emptyDir sizeLimit to 5GiB. With this setting in pod  ...
```
I'm not sure of the math here. I could be wrong, conceptually, but if the first container requests 2 GiB & the second container requests 6 GiB shouldn't the total requested be 8 GiB? We either need to change the example or the math.

```suggestion
In the following example, the Pod has two containers. The first container has a request of 2GiB of local ephemeral storage and a limit of 5GiB. The second container requests 6GiB of local storage. Therefore, the Pod requests a total of 10GiB (8GiB+2GiB) of local ephemeral storage and enforces a limit of 6GiB of local ephemeral storage. It also sets emptyDir sizeLimit to 5GiB. With this setting in pod spec, it will affect how scheulder makes a decision on scheduling pods and also how kubelet evict pods.
```

divya-mohan0209 reviewed (commented):
Some suggestions, hope you find them useful.

xing-yang commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+Similar to CPU and memory, admin could use LimitRange to set default container’s local storage request/limit, and/or minimum/maximum resource constraints for a namespace.
+
+Also, ephemeral-storage may be specified to reserve for kubelet or system. example, --system-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000] --kube-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000]
```
How does this "reserve" work together with those requests/limits mentioned above?  Can you add some explanation in the blog?

jingxu97 replied:
added more information there. PTAL

xing-yang commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+This project, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together.
+
+We offer a huge thank you to all the contributors in Kubernetes Storage SIG and CSI community who helped review the design and implementation of the project, including but not limited to the following:</p><ul><li>Benjamin Elder (<a href=https://github.com/BenTheElder>BenTheElder</a>)</li><li>Michelle Au (<a href=https://github.com/msau42>msau42</a>)</li><li>Tim Hockin (<a href=https://github.com/ ...
```
Add a hyper link to https://github.com/kubernetes/community/tree/master/sig-storage?

NitishKumar06 commented:
@jingxu97 Please go through the changes that @divya-mohan0209 and @xing-yang mentioned.

Sea-n commented:
Hey @jingxu97,

It's been a while since the last update, would you like to address issues in the previous comments so that we can move forward?

xing-yang commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+---
+layout: blog
+title: "Kubernetes 1.25: Local Storage Capacity Isolation Reaches GA"
+date: 2022-08-15
```
Change to 2022-09-19 which is the publish date

sftim replied:
```suggestion
date: 2022-09-19
```
(and also please change the path of the file to match)

NitishKumar06 replied:
@xing-yang Are you active for this issue?

xing-yang replied:
Yes.  @jingxu97 Can you please address the review comments?

jingxu97 replied:
done, also changed the file name.

sftim commented on content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,94 @@
+
+This project, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together.
+
+We offer a huge thank you to all the contributors in Kubernetes Storage SIG and CSI community who helped review the design and implementation of the project, including but not limited to the following:</p><ul><li>Benjamin Elder (<a href=https://github.com/BenTheElder>BenTheElder</a>)</li><li>Michelle Au (<a href=https://github.com/msau42>msau42</a>)</li><li>Tim Hockin (<a href=https://github.com/ ...
```
Eliding the “to” is common in US English. I think that's fine for a blog article.

sftim reviewed (commented):
@jingxu97 we'd love to publish this article, but we'd need updates to this article as soon as you can make them.

jingxu97 commented:
> @jingxu97 we'd love to publish this article, but we'd need updates to this article as soon as you can make them.

@sftim, very sorry for the delay. I addressed the comments, PTAL. Thank you!

xing-yang commented on content/en/blog/_posts/2022-09-19-local-storage-capacity-isolation-ga.md:
```diff
@@ -0,0 +1,107 @@
+* `spec.containers[].resources.limits.ephemeral-storage`
+* `spec.containers[].resources.requests.ephemeral-storage`
+
+In the following example, the Pod has two containers. The first container has a request of 8GiB of local ephemeral storage and a limit of 12GiB. The second container requests 2GiB of local storage, but no limit setting. Therefore, the Pod requests a total of 10GiB (8GiB+2GiB) of local ephemeral storage and enforces a limit of 6GiB of local ephemeral storage. It also sets emptyDir sizeLimit to 5Gi ...
```
"enforces a limit of 6GiB of local ephemeral storage"?  Why 6GiB?  Should the limit be 12GiB (12+0, the first container sets limit to 12 but the 2nd container has no limit setting)?

xing-yang commented:
/lgtm

reylejano commented:
/approve
